src/ddmd/attrib.d

```diff
@@ -1352,6 +1384,169 @@ extern (C++) final class StaticIfDeclaration : ConditionalDeclaration
+    }
+
+    override final bool oneMember(Dsymbol* ps, Identifier ident)
+    {
+        if (cached)
+        {
+            return super.oneMember(ps, ident);
+        }
+        *ps = null;
+        return false;
+    }
+
+    override Dsymbols* include(Scope* sc, ScopeDsymbol sds)
+    {
+        if (cached)
+        {
+            return cache;
+        }
+        sfe.prepare(_scope);
+        if (!sfe.ready())
+        {
+            return null; // TODO: ok?
+        }
+
+        import ddmd.statementsem: makeTupleForeach;
+        Dsymbols* d = makeTupleForeach!(true,true)(_scope, sfe.aggrfe, decl, sfe.needExpansion);
+        if (d && !addisdone)
+        {
+            // Add members lazily.
+            for (size_t i = 0; i < d.dim; i++)
+            {
+                Dsymbol s = (*d)[i];
+                s.addMember(_scope, scopesym);
+            }
+            // Set the member scopes lazily.
+            for (size_t i = 0; i < d.dim; i++)
+            {
+                Dsymbol s = (*d)[i];
+                s.setScope(_scope);
+            }
+            addisdone = true;
+        }
+        cached = true;
+        cache = d;
+        return d;
+    }
+
+    override void addMember(Scope* sc, ScopeDsymbol sds)
+    {
+        this.scopesym = sds;
+    }
+
+    override final void addComment(const(char)* comment)
+    {
+        // do nothing
+        // change this to give a semantics to documentation comments on static foreach declarations
+    }
+
+    override void setScope(Scope* sc)
+    {
+        // do not evaluate condition before semantic pass
+        // But do set the scope, in case we need it for forward referencing
+        Dsymbol.setScope(sc);
+    }
+
+    override void importAll(Scope* sc)
+    {
+        // do not evaluate aggregate before semantic pass
+    }
+
+    override void semantic(Scope* sc)
+    {
+        AttribDeclaration.semantic(sc);
+    }
+
+    override const(char)* kind() const
+    {
+        return "static foreach";
+    }
+
+    override void accept(Visitor v)
+    {
+        v.visit(this);
+    }
+}
+
+/***********************************************************
+ * Collection of declarations that stores foreach index variables in a local scope.
+ * Symbols declared within are inserted into another scope, like:
+ *
+ *      static foreach (i; 0 .. 10) // loop variables for different indices do not conflict.
+ *      { // this body is expanded into 10 ForwardingAttribDeclarations, where i has storage class STClocal
+ *          mixin("enum x" ~ to!string(i) ~ " = i"); // ok, can access current loop variable
+ *      }
+ *
+ *      static foreach (i; 0.. 10)
+ *      {
+ *          pragma(msg, mixin("x" ~ to!string(i))); // ok, all 10 symbols are visible
+ *      }
+ *
+ *      static assert (!is(typeof(i))); // loop index variable is not visible outside of the static foreach loop
+ */
+
+extern(C++) final class ForwardingAttribDeclaration: AttribDeclaration
```

Comment:
It is called forwarding because symbols are forwarded to a different scope upon insertion. The foreach index variables have storage class `STClocal` which suppresses forwarding I will add this to the documentation comment.